### PR TITLE
DDB SQS deduping implementation

### PR DESCRIPTION
**Pull request by JDragovichAlertLogic**
### Problem Description
This attempts to filter out duplicate messages from the sqs queue.

it is a sketch and is meant to be minimally invasive to existing collectors. It is based on this medium article.

https://medium.com/avmconsulting-blog/deduplicating-amazon-sqs-messages-dc114d1e6545

### Solution Description
The collector now tracks states in a DDB table. before doing anything. teh collector checks to see if the state is already in the table, then checks to see if it is currently being processed by another invocation. If not, it proceeds with collection.

Once the collector is done and sent its logs to ingest. It marks the state as complete in the table.

### CODE REVIEW

File: paws_collector.js [hunk 1]
```diff
@@ -250,7 +253,7 @@ class PawsCollector extends AlAwsCollector {
         });
     };
 
-    handleEvent(event) {
+    handleEvent(event, asyncCallback) {
```
Comment (kkuzmin) [inline paws_collector.js:256]: `asyncCallback` is not used.

File: paws_collector.js [hunk 2]
```diff
@@ -263,6 +266,83 @@ class PawsCollector extends AlAwsCollector {
...
+        const DDB = new AWS.DynamoDB();
+
+        const newRecord = {
+            Item: {
+                CollectorId: {S: collector._collectorId},
+                MessageId: {S: stateSqsMsg.MD5OfBody},
+                Updated: {N: Date.now().toString()},
+                Status: {S: Status},
+            },
+            TableName: tableName
```
Comment (kkuzmin) [inline paws_collector.js:336]: I can't find `tableName ` defiled in this function.

File: paws_collector.js [hunk 3]
```diff
@@ -263,6 +266,83 @@ class PawsCollector extends AlAwsCollector {
...
+        }
```
Comment (kkuzmin) [inline paws_collector.js:337]: please add semicolon here.

File: paws_collector.js [hunk 4]
```diff
@@ -263,6 +266,83 @@ class PawsCollector extends AlAwsCollector {
...
+        DDB.putItem(newRecord, (err) => {
+            if(err){
+                // figure out how to handle ddb write errors
```
Comment (kkuzmin) [inline paws_collector.js:340]: we need to just fail the invocation so that SQS message is not consumed and retried.

File: paws_collector.js [hunk 5]
```diff
@@ -263,6 +266,83 @@ class PawsCollector extends AlAwsCollector {
...
+            TableName: tableName,
+            ConsistentRead: true
+        }
+
+        const getItemPromise = DDB.getItem(params).promise();
+
+        getItemPromise.then(data => {
+            // if the item is alread there, try and see if it is a duplicate
+            if (data.Item) {
+                // not sure if its right to call collector.done here
```
Comment (kkuzmin) [inline paws_collector.js:291]: simply return an error here. `async.waterfall()` from `handlePoll` will handle this error for you.

File: paws_collector.js (hunk 4 above)
Comment (kkuzmin) [reply paws_collector.js:340]: please return an error and `async.waterfall()` from `handleEvent` will handle it for you.

File: paws_collector.js [hunk 6]
```diff
@@ -263,6 +266,83 @@ class PawsCollector extends AlAwsCollector {
...
+                        Updated: {N: Date.now().toString()},
+                        Status: {S: STATE_RECORD_INCOMPLETE},
+                        // setting DDB time to life. This is the same as the sqs queue message retention
+                        ExpireDate: {N: moment().add(14, 'days').unix().toString()}
+                    },
+                    TableName: tableName
+                }
+                DDB.putItem(newRecord, (err) => {
+                    if(err){
+                        // TODO:figure out how to handle ddb write errors. Do they crash the collector?
```
Comment (kkuzmin) [inline paws_collector.js:315]: please return an error and `async.waterfall()` from `handleEvent` will handle it for you.

**Review by kkuzmin — CHANGES_REQUESTED** [review]
thanks @JDragovichAlertLogic 
The current solution stores a DDB record per every SQS message received by collector. In general this is good for simplicity and traceability. And I think it is ok to try this out.
I was thinking about more complicated solution where we store just one DDB record (like, last collected interval from state) per each collector/stream and collector will check it before handling the SQS. I think we can go back to this solution if we see increased DDB running costs for large collectors.

**Comment by JDragovichAlertLogic** [comment]
test it

File: paws_collector.js [hunk 7]
```diff
@@ -263,6 +266,93 @@ class PawsCollector extends AlAwsCollector {
...
+    // check the status of the state in DDB to avoid duplication
+    // there is a lot of logic here. not sur eif there is a better way of deduping. trying for an MVP
+    checkStateSqsMessage(stateSqsMsg, asyncCallback) {
+        const collector = this;
+        const DDB = new AWS.DynamoDB();
+        const tableName = process.env.pawsDDBTableName;
+
+        const params = {
+            Key: {
+                "CollectorId": {S: collector._collectorId},
```
Comment (kkuzmin) [inline paws_collector.js:279]: please add CID into the table because we'll be having collectors from multiple CIDs

File: paws_collector.js [hunk 8]
```diff
@@ -263,6 +266,93 @@ class PawsCollector extends AlAwsCollector {
...
+        }
+
+        const getItemPromise = DDB.getItem(params).promise();
+
+        getItemPromise.then(data => {
+            // if the item is alread there, try and see if it is a duplicate
+            if (data.Item) {
+                const Item = data.Item;
+                if(Item.Status.S === STATE_RECORD_INCOMPLETE && moment().unix() - parseInt(Item.Updated.N) < 900) {
+                    console.log(`Duplicate state: ${Item.MessageId.S}, already in progress. skipping`);
```
Comment (kkuzmin) [inline paws_collector.js:293]: Please use `console.info()`

File: paws_collector.js [hunk 9]
```diff
@@ -263,6 +266,93 @@ class PawsCollector extends AlAwsCollector {
...
+                    console.log(`Duplicate state: ${Item.MessageId.S}, already processed. skipping`);
+                    return collector.done();
+                } else {
+                    return collector.updateStateDBEntry(stateSqsMsg, STATE_RECORD_INCOMPLETE, asyncCallback);
+                }
+            // otherwise, put a new item in ddb.
+            } else {
+                const newRecord = {
+                    Item: {
+                        CollectorId: {S: collector._collectorId},
```
Comment (kkuzmin) [inline paws_collector.js:305]: please add Cid.

File: paws_collector.js [hunk 10]
```diff
@@ -263,6 +266,93 @@ class PawsCollector extends AlAwsCollector {
...
+                });
+            }
+            
+        }).catch(asyncCallback)
+    }
+
+    updateStateDBEntry(stateSqsMsg, Status, asyncCallback) {
+        const collector = this;
+        const DDB = new AWS.DynamoDB();
+        const tableName = process.env.pawsDDBTableName;
```
Comment (kkuzmin) [inline paws_collector.js:329]: please add DDB table name into Paws [class properties](https://github.com/alertlogic/paws-collector/pull/147/files#diff-0334275db1035801ee0894723546b59aR111) and use it here

File: paws_collector.js [hunk 11]
```diff
@@ -303,6 +396,9 @@ class PawsCollector extends AlAwsCollector {
             },
             function(privCollectorState, nextInvocationTimeout, asyncCallback) {
                 return collector._storeCollectionState(pawsState, privCollectorState, nextInvocationTimeout, asyncCallback);
+            },
+            function(_results, asyncCallback) {
+                return collector.updateStateDBEntry(stateSqsMsg, STATE_RECORD_COMPLETE, asyncCallback);
```
Comment (kkuzmin) [inline paws_collector.js:401]: I think we can move `updateStateDBEntry()` into the function above and use `async.parallel` to run them independently. Ideas?

File: paws_collector.js [hunk 12]
```diff
@@ -263,6 +266,93 @@ class PawsCollector extends AlAwsCollector {
...
+            ConsistentRead: true
+        }
+
+        const getItemPromise = DDB.getItem(params).promise();
+
+        getItemPromise.then(data => {
+            // if the item is alread there, try and see if it is a duplicate
+            if (data.Item) {
+                const Item = data.Item;
+                if(Item.Status.S === STATE_RECORD_INCOMPLETE && moment().unix() - parseInt(Item.Updated.N) < 900) {
```
Comment (kkuzmin) [inline paws_collector.js:292]: Please add a comment to clarify `moment().unix() - parseInt(Item.Updated.N) < 900`.
Also please move `900` into some global constant.

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: paws_collector.js (hunk 11 above)
Comment (JDragovichAlertLogic) [reply paws_collector.js:404]: We need to make sure the SqS push succeeds before we update the DDB state.

If we run them in paralelle, the DDB call might succeed where the SQS push might fail. If that happens, the retry of the state will be marked as a duplicate, and stop the collector.

File: cfn/paws-collector.template [hunk 13]
```diff
@@ -12,6 +12,11 @@
             "Type": "String",
             "Default": ""
         },
+        "DdbTableName" :{
```
Comment (kkuzmin) [inline cfn/paws-collector.template:15]: Lets name this param as `ExistingDDBTableName`. to make it clearer.

File: cfn/paws-collector.template [hunk 14]
```diff
@@ -125,7 +130,41 @@
...
+                        "KeyType" : "HASH"
+                    },
+                    {
+                        "AttributeName" : "MessageId",
+                        "KeyType" : "RANGE"
+                    }
+                ],
+                "TimeToLiveSpecification" : {
+                    "AttributeName" : "ExpireDate",
+                    "Enabled" : true
```
Comment (kkuzmin) [inline cfn/paws-collector.template:164]: where do we set expiration timeout to 2 weeks?

**Review by kkuzmin — CHANGES_REQUESTED** [review]

File: paws_collector.js [hunk 15]
```diff
@@ -263,6 +268,98 @@ class PawsCollector extends AlAwsCollector {
...
+            } else {
+                const newRecord = {
+                    Item: {
+                        CollectorId: {S: collector._collectorId},
+                        MessageId: {S: stateSqsMsg.md5OfBody},
+                        Updated: {N: moment().unix().toString()},
+                        Cid: {S: collector.cid ? collector.cid : 'none'},
+                        Status: {S: STATE_RECORD_INCOMPLETE},
+                        // setting DDB time to life. This is the same as the sqs queue message retention
+                        ExpireDate: {N: moment().add(14, 'days').unix().toString()}
```
Comment (kkuzmin) [inline paws_collector.js:314]: Please declares 2 weeks constant on top

File: cfn/paws-collector.template [hunk 16]
```diff
@@ -125,7 +130,41 @@
             "Default": ""
         }
     },
+    "Conditions":{
+        "createDdbTable": {"Fn::Equals" : [{"Ref" :"DdbTableName" }, ""]}
```
Comment (kkuzmin) [inline cfn/paws-collector.template:134]: should be `ExistingDDBTableName `

File: cfn/paws-collector.template [hunk 17]
```diff
@@ -571,6 +610,13 @@
                   "DD_KMS_API_KEY": {
                       "Ref": "DatadogKMSKey"
                   },
+                  "pawsDDBTableName": {
+                      "Fn::If": [
+                        "createDdbTable",
+                        {"Ref": "DdupeDdbTable"},
+                        {"Ref": "DdbTableName"}
```
Comment (kkuzmin) [inline cfn/paws-collector.template:617]: should be `DdbTableName `

File: cfn/paws-collector.template [hunk 18]
```diff
@@ -753,6 +799,33 @@
...
+                          [
+                             "arn:aws:dynamodb:",
+                             { "Ref" : "AWS::Region" }, ":",
+                             { "Ref" : "AWS::AccountId"}, ":",
+                             "table/",
+                             {
+                                "Fn::If": [
+                                    "createDdbTable",
+                                    {"Ref": "DdupeDdbTable"},
+                                    {"Ref": "DdbTableName"}
```
Comment (kkuzmin) [inline cfn/paws-collector.template:822]: shoudl be `DdbTableName`

File: paws_collector.js [hunk 19]
```diff
@@ -106,6 +111,7 @@ class PawsCollector extends AlAwsCollector {
         this._pawsEndpoint = process.env.paws_endpoint
         this._pawsDomainEndpoint = endpointDomain;
         this._pawsHttpsEndpoint = 'https://' + endpointDomain;
+        this.ddbTableName = process.env.pawsDDBTableName;
```
Comment (kkuzmin) [inline paws_collector.js:114]: lets' use `_pawsDdbTableName`, to keep it consistent with the rest of properties

**Review by kkuzmin — APPROVED** [review]

**Pull request merged**